pkg/datapath/state/device.go

```diff
@@ -0,0 +1,137 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package state
+
+import (
+	"fmt"
+	"net"
+	"net/netip"
+
+	"github.com/hashicorp/go-memdb"
+	"github.com/vishvananda/netlink"
+	"golang.org/x/exp/slices"
+
+	"github.com/cilium/cilium/pkg/statedb"
+)
+
+var deviceTableSchema = &memdb.TableSchema{
+	Name: "devices",
+	Indexes: map[string]*memdb.IndexSchema{
+		"id": {
+			Name:         "id",
+			AllowMissing: false,
+			Unique:       true,
+			Indexer:      &memdb.IntFieldIndex{Field: "Index"},
+		},
+		"name": {
+			Name:         "name",
+			AllowMissing: false,
+			Unique:       true,
+			Indexer:      &memdb.StringFieldIndex{Field: "Name"},
+		},
+	},
+}
+
+// Device is a local network device along with addresses associated with it.
+type Device struct {
+	Index  int             // Interface index (primary key)
+	Name   string          // Interface name (e.g. eth0)
+	Viable bool            // If true this device can be used by Cilium
+	Addrs  []DeviceAddress // Addresses assigned to the device
+	Link   netlink.Link    // The underlying link device
+}
+
+func (d *Device) String() string {
+	return fmt.Sprintf("Device{Index:%d, Name:%s, len(Addrs):%d}", d.Index, d.Name, len(d.Addrs))
+}
+
+func (d *Device) DeepCopy() *Device {
+	copy := *d
+	copy.Addrs = slices.Clone(d.Addrs)
+	return &copy
+}
+
+func (d *Device) HasIP(ip net.IP) bool {
+	for _, addr := range d.Addrs {
+		if addr.AsIP().Equal(ip) {
+			return true
+		}
+	}
+	return false
+}
+
+type NodePortAddr struct {
+	IPv4 net.IP
+	IPv6 net.IP
+}
+
+// NodePortAddr returns the NodePort frontend addresses chosen
+// from this device.
+func (dev *Device) NodePortAddr() NodePortAddr {
```

Comment:
Do we actually need this function to be NodePort-specific? I guess in the future we will support multiple NodePort IP addrs per netdev, so maybe we could use a generic `IPAddrs()` method instead?

Reply:
To elaborate more - I don't want to leak any LB details (in this case NodePort) into this relatively generic package.

Reply:
Since this method is about picking the single nodeport frontend for the device I would keep it NodePort-specific. However, you're right it shouldn't be here. I'll move it to a more appropriate place. In the future this would go away completely once service load-balancing is refactored to support multiple frontends per device.